unit/unit_servicemanager_aidl4/unit_servicemanager_aidl4.c

```diff
@@ -0,0 +1,488 @@
+            remote_obj = g_hash_table_lookup(self->objects, str);
+            if (remote_obj) {
+                GBinderWriter writer;
+
+                GDEBUG("Found name '%s' => %p", str, remote_obj);
+                gbinder_local_reply_init_writer(reply, &writer);
+                gbinder_writer_append_int32(&writer, UNDECLARED);
+                gbinder_writer_append_remote_object(&writer, remote_obj);
+            } else {
+                GDEBUG("Name '%s' not found", str);
+                gbinder_local_reply_append_int32(reply, GBINDER_STATUS_OK);
+            }
+            g_free(str);
+        }
+        break;
+    case ADD_SERVICE_TRANSACTION:
+        gbinder_remote_request_init_reader(req, &reader);
+        str = gbinder_reader_read_string16(&reader);
+        remote_obj = gbinder_reader_read_object(&reader);
+        gbinder_reader_read_uint32(&reader, (guint32*)&category);
+        if (str && remote_obj &&
+            category.level == SYSTEM &&
+            category.version == 1 &&
+            gbinder_reader_read_uint32(&reader, &allow_isolated) &&
+            gbinder_reader_read_uint32(&reader, &dumpsys_priority)) {
+            GDEBUG("Adding '%s'", str);
+            g_hash_table_replace(self->objects, str, remote_obj);
+            remote_obj = NULL;
+            str = NULL;
+            reply = gbinder_local_object_new_reply(obj);
+            *status = GBINDER_STATUS_OK;
+        }
+        g_free(str);
+        gbinder_remote_object_unref(remote_obj);
+        break;
+    case LIST_SERVICES_TRANSACTION:
+        gbinder_remote_request_init_reader(req, &reader);
+        if (gbinder_reader_read_uint32(&reader, &dumpsys_priority)) {
+            if (g_hash_table_size(self->objects) == 1) {
+                GList* keys = g_hash_table_get_keys(self->objects);
+                GList* l = g_list_nth(keys, 0);
+                gint32 srv_size = 1;
+                GBinderWriter writer;
+
+                reply = gbinder_local_object_new_reply(obj);
+                gbinder_local_reply_init_writer(reply, &writer);
+                gbinder_writer_append_int32(&writer, GBINDER_STATUS_OK);
+                gbinder_writer_append_int32(&writer, srv_size);
+                gbinder_writer_append_string16(&writer, l->data);
+                g_list_free(keys);
+                *status = GBINDER_STATUS_OK;
+            } else {
+                GDEBUG("Incorrect number of services %u",
+                    g_hash_table_size(self->objects));
+            }
+        }
+        break;
+    default:
+        GDEBUG("Unhandled command %u", code);
+        break;
+    }
+    return reply;
+}
+
+static
+ServiceManagerAidl4*
+servicemanager_aidl4_new(
+    const char* dev,
+    gboolean handle_on_looper_thread)
+{
+    ServiceManagerAidl4* self = g_object_new(SERVICE_MANAGER_AIDL4_TYPE, NULL);
+    GBinderLocalObject* obj = GBINDER_LOCAL_OBJECT(self);
+    GBinderIpc* ipc = gbinder_ipc_new(dev);
```

Comment:
`gbinder_ipc_new` is an internal API and it recently changed:

```
unit_servicemanager_aidl4.c: In function ‘servicemanager_aidl4_new’:
unit_servicemanager_aidl4.c:211:23: error: too few arguments to function ‘gbinder_ipc_new’
  211 |     GBinderIpc* ipc = gbinder_ipc_new(dev);
      |                       ^~~~~~~~~~~~~~~
In file included from unit_servicemanager_aidl4.c:37:
../../src/gbinder_ipc.h:102:1: note: declared here
  102 | gbinder_ipc_new(
      | ^~~~~~~~~~~~~~~
unit_servicemanager_aidl4.c: In function ‘test_context_init’:
unit_servicemanager_aidl4.c:341:11: error: too few arguments to function ‘gbinder_ipc_new’
  341 |     ipc = gbinder_ipc_new(dev);
      |           ^~~~~~~~~~~~~~~
In file included from unit_servicemanager_aidl4.c:37:
../../src/gbinder_ipc.h:102:1: note: declared here
  102 | gbinder_ipc_new(
      | ^~~~~~~~~~~~~~~
make[2]: *** [../common/Makefile:180: build/debug/unit_servicemanager_aidl4.o] Error 1
make[2]: Leaving directory '/home/slava/projects/jolla/libgbinder/unit/unit_servicemanager_aidl4'
make[1]: *** [Makefile:28: test] Error 2
make[1]: Leaving directory '/home/slava/projects/jolla/libgbinder/unit'
make: *** [Makefile:249: test] Error 2
```

Reply:
Thanks, fixed.